File: src/providers/openai-realtime/rules/verify-deprecated-session-fields.ts
import { findProperty } from '../utils.js';

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: "OpenAI Realtime session.update payloads should not rely on an unverified 'temperature' field",
      category: 'correctness',
      docsUrl: 'https://developers.openai.com/api/docs/api-reference/realtime-sessions',
      rationale:
        "Two independent fetches of the current Realtime sessions API reference did not surface 'temperature' as a documented session field. Sending an undocumented field is typically harmless (silently ignored), but code that depends on an assumed constraint (e.g. 'a hard floor of 0.6') for behavior like deterministic output is relying on a claim that is no longer traceable to any current, citable source.",
      recommended: true,
    },
    messages: {
      unverifiedTemperatureField:
        "This session.update payload sets 'temperature', a field not documented in the current GA Realtime sessions schema.",
    },
    schema: [],
  },
  create(context: any) {
    return {
      ObjectExpression(node: any) {
        const typeProp = findProperty(node, 'type');
        const typeValue = typeProp?.value;
        const isSessionUpdate =
          typeValue?.type === 'Literal' && typeof typeValue.value === 'string' && typeValue.value === 'session.update';
        if (!isSessionUpdate) return;

        const sessionProp = findProperty(node, 'session');
        if (sessionProp?.value?.type !== 'ObjectExpression') return;

        const temperatureProp = findProperty(sessionProp.value, 'temperature');
        if (!temperatureProp) return;

        context.report({ node: temperatureProp, messageId: 'unverifiedTemperatureField' });
      },
    };
  },
};

export const openaiRealtimeVerifyDeprecatedSessionFieldsRule = rule;

File: src/providers/openai-realtime/rules/python/verify_deprecated_session_fields.py
"""openai-realtime-verify-deprecated-session-fields

Parity with the JS rule: a `session.update` payload setting `temperature`
relies on a field not documented in the current GA Realtime sessions schema.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

from ast_utils import loc

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import dict_get, dict_get_pair  # noqa: E402

RULE_KEY = "openai-realtime-verify-deprecated-session-fields"


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []

    for node in ast.walk(tree):
        if not isinstance(node, ast.Dict):
            continue
        type_val = dict_get(node, "type")
        if not (isinstance(type_val, ast.Constant) and type_val.value == "session.update"):
            continue

        session_val = dict_get(node, "session")
        if not isinstance(session_val, ast.Dict):
            continue

        pair = dict_get_pair(session_val, "temperature")
        if pair is None:
            continue

        key_node, _ = pair
        line, col, _, _ = loc(key_node)
        out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out
